Description: Added "dumparray" "setlocal" variable
Author: vir@ctm.ru
---
If this variable is set, ARRAY responses will be converted into tab-separated text, allowing external scripts to process database query responses.
Index: yate/modules/extmodule.cpp
===================================================================
--- yate.orig/modules/extmodule.cpp	2016-01-15 15:06:17.000000000 +0300
+++ yate/modules/extmodule.cpp	2016-01-15 15:44:07.000000000 +0300
@@ -210,6 +210,7 @@
 class ExtModReceiver : public MessageReceiver, public Mutex
 {
     friend class MsgWatcher;
+    friend class ExtMessage;
 public:
     enum {
 	RoleUnknown,
@@ -283,6 +284,7 @@
     ObjList m_relays;
     String m_trackName;
     String m_reason;
+    bool m_dumparray;
 };
 
 class ExtThread : public Thread
@@ -619,10 +621,44 @@
     return (m_msg.decode(s,m_ret,m_id) == -2);
 }
 
+static String escape(const char* str)
+{
+    String s;
+    if (TelEngine::null(str))
+	return s;
+    char c;
+    while ((c=*str++)) {
+	if (c == '\\' || c == '\t' || c == '\n')
+	    s += "\\";
+	s += c;
+    }
+    return s;
+}
+
+static String dumpArray(const Array *a)
+{
+    String r;
+    for (int j=0; j < a->getRows(); j++) {
+	for (int i=0; i<a->getColumns(); i++) {
+	    if(i)
+		r << "\t";
+	    const String* s = YOBJECT(String, a->get(i, j));
+	    if (s)
+		r << escape(*s);
+	}
+	r << "\n";
+    }
+    return r;
+}
 
 ExtMessage::~ExtMessage()
 {
     if (m_receiver) {
+	if(m_receiver->m_dumparray && userData()) {
+	    Array* a = static_cast<Array*>(userObject(YATOM("Array")));
+	    if(a)
+		retValue() = dumpArray(a);
+	}
 	m_receiver->returnMsg(this,m_id,m_accepted);
 	m_receiver->unuse();
     }
@@ -778,7 +814,8 @@
       m_chan(chan), m_watcher(0),
       m_selfWatch(false), m_reenter(false), m_setdata(true), m_writing(false),
       m_timeout(s_timeout), m_timebomb(s_timebomb), m_restart(false), m_scripted(false),
-      m_buffer(0,DEF_INCOMING_LINE), m_script(script), m_args(args), m_trackName(s_trackName)
+      m_buffer(0,DEF_INCOMING_LINE), m_script(script), m_args(args), m_trackName(s_trackName),
+      m_dumparray(false)
 {
     Debug(DebugAll,"ExtModReceiver::ExtModReceiver(\"%s\",\"%s\") [%p]",script,args,this);
     m_script.trimBlanks();
@@ -1569,6 +1606,11 @@
 		ok = val.null();
 		val = Engine::runId();
 	    }
+	    else if (id == "dumparray") {
+		m_dumparray = val.toBoolean(m_dumparray);
+		val = m_dumparray;
+		ok = true;
+	    }
 	    DDebug("ExtModReceiver",DebugAll,"Set '%s'='%s' %s",
 		id.c_str(),val.c_str(),ok ? "ok" : "failed");
 	    String out("%%<setlocal:");
